### ISSUE TITLE
Investigate basis issues in BASEX

### ISSUE THREAD

**Issue by MikhailRyazanov** [PyAbel/PyAbel#230]
There are several problems in the `basex` method related to the underlying basis sets.
1. `nbf` ≠ `n` is currently not implemented. The utility of `nbf` < `n` or `nbf` > `n` is questionable, but the original BASEX had this capability, and it worked.
2. The mystery of the `MAGIC_NUMBER`, see #4. Currently, it is some hard-coded heuristic scaling factor, although it should have some reasonable mathematical origin and an analytical expression (probably, with basis-width dependence).
3. Something strange happens near _r_ = 0. See https://github.com/PyAbel/PyAbel/issues/225#issuecomment-422992673:
![profile8](https://user-images.githubusercontent.com/10932229/45787421-95524000-bcb8-11e8-96fb-a9b826ca3029.png)

**Comment by MikhailRyazanov** [comment]
Some thoughts about point 3.

The image basis functions look like this:
![basis](https://user-images.githubusercontent.com/12854133/46057071-489cb800-c143-11e8-86b7-1da68e56f832.png)
Colored lines show functions _ρ<sub>k</sub>_(_r_) centered at _r_ = _k_, and black is their sum (over _k_ < 10). Dashed are the actual functions, and the solid lines represent what is in the basis matrix (they are really just points at integer _r_ values).

Two simple observations: the _k_ = 0 function is broader than all other, and all other functions are zero at _r_ = 0.

Thus the sum at _r_ = 0 is lower than anywhere else (no contributions from _k_ > 0), and at small _r_ > 0 it is higher than anywhere else (the tail of _k_ = 0).

In general, this looks quite similar to the behavior seen in the example above. The problem is that in principle these basis functions are still linearly independent and cover the full _r_ range, so to work correctly they do not need to be normalized (or to be orthogonal, or whatever else), as long as their projections are correct.

The projections in the article apparently have a wrong factor: it is "2" in (15), but my calculations show that (at least for _k_ = 0) it should be π<sup>1/2</sup>. By the way, 2/π<sup>1/2</sup> ≈ 1.128, which is relatively close but not equal to `MAGIC_NUMBER = 1.112...`. However, that "wrong" factor is used for all projections, so it should produce wrong normalization, but not the observed artifact. Unless the projected-basis computation has errors?

**Comment by MikhailRyazanov** [comment]
Here are the basis projections:
![projections](https://user-images.githubusercontent.com/12854133/46060634-8dc8e600-c153-11e8-8f78-da065936d580.png)
As before, dashed are _χ<sub>k</sub>_(_x_), calculated directly using (15) in the article, solid are the matrix values.

So at least the implementation seems to be consistent with what is written in the article. I do not know whether the integrals there are correct ([obviously?](https://www.reddit.com/r/math/comments/5kjjfb/the_story_of_the_professor_who_said_it_was_trivial/dbojmgh/)).

But the plots show quite clearly that these functions are too narrow for this sampling frequency and, unlike the _ρ_ functions, for _k_ > 0 they peak _between_ the sampling points, so the projected-basis matrix reproduces them rather poorly. Is this the cause of the problem? If yes, then broader basis sets should be less affected.

**Comment by MikhailRyazanov** [comment]
An attempt to correct the _r_ = 0 problem.

Modifying the _k_ = 0 basis function such that the sum is (more) flat:
![basis_corrected](https://user-images.githubusercontent.com/12854133/46064444-ebaefb00-c15e-11e8-8b6b-dcf644b40c9d.png)
This is a taller narrower Gaussian to compensate _r_ = 0, minus a little bit of a broad Gaussian to compensate the tail. Namely:
```python
    An = 1.27135
    wn = 0.76
    Ab = 0.04
    wb = 3.0
    Mc[:, 0] = (An + Ab) * np.exp(-R2_h/wn**2) - \
                      Ab * np.exp(-R2_h/wb**2)
```

Using its projection
```python
    M[:, 0] = 2 * (wn * (An + Ab) * np.exp(-R2_h/wn**2) -
                   wb *        Ab * np.exp(-R2_h/wb**2))
```
does not eliminate the artifact, but somewhat decreases its magnitude.

So I've done some manual tweaking of the projection samples to minimize the oscillations near _r_ = 0. Here are the "corrected" projections:
![projections_corrected](https://user-images.githubusercontent.com/12854133/46064758-d1295180-c15f-11e8-8a0c-37a1a60d444f.png)
(Only the _k_ = 0 function was changed. Notice that its tweaked samples slightly deviate from the projected function.)

Since this affects only several elements in the basis matrices, I've [added](https://github.com/MikhailRyazanov/PyAbel/tree/basis) a `bs_correction` parameter that does this basis modification in `get_bs_basex_cached`. It does not require recomputing the basis matrices (so current disk caches remain valid) and can be used/unused in each `basex_transform` call arbitrarily (but will trigger basis reloading when changed).

Repeating the test from https://github.com/PyAbel/PyAbel/issues/225#issuecomment-422992673 using
```python
b = abel.basex.basex_transform(f.abel, dr=f.dr, bs_correction=True)
```
gives the following result:
![profile8](https://user-images.githubusercontent.com/12854133/46065248-1732e500-c161-11e8-938f-b38e8848ac51.png)
Looks much better, although still not perfect.

@rth, could you please try this modification with various distributions/images where you saw these problems to check how does it work? I tested it only with Gaussians (using various widths and basis sizes), and it always was better, but who knows.

**Comment by DanHickstein** [comment]
Thanks @MikhailRyazanov for opening this important issue!

Is your correction to the basis set entirely empirical?

**Comment by stggh** [comment]
> @stggh, could you please try this modification

This is tricky. In general the *dirty hack!* improves the mean-square-error (top mse = `basex`, lower = `3pt`) because it tidies up the first 2 bad points:
![testbasex101false](https://user-images.githubusercontent.com/10932229/46119107-eaf49280-c24c-11e8-9f0a-12edad14b3c5.png)
![testbasex101true](https://user-images.githubusercontent.com/10932229/46119109-ecbe5600-c24c-11e8-92ab-10011746ec9f.png)

Zooming in shows that there is more to the story:
![testbasex101falsezoom](https://user-images.githubusercontent.com/10932229/46119175-34dd7880-c24d-11e8-8c3f-303595eee7a4.png)
![testbasex101truezoom](https://user-images.githubusercontent.com/10932229/46119184-360ea580-c24d-11e8-8567-064fbd8d7f35.png)

It is clearer for more datapoints:
![testbasex501falsezoom](https://user-images.githubusercontent.com/10932229/46119227-69e9cb00-c24d-11e8-844d-f61d720119e9.png)
![testbasex501truezoom](https://user-images.githubusercontent.com/10932229/46119228-6b1af800-c24d-11e8-9172-4bbfbddaf371.png)

`bs_correction = True` does a good job tidying the first few points of the transform, but it is not correcting the underlying physics. 

[code gist](https://gist.github.com/stggh/1f4419807aa201696e1f6fe8100b7880)

**Comment by MikhailRyazanov** [comment]
> ... is not correcting the underlying physics.

Mathematics. :–)

I have some ideas to try.
1. A broad basis. For _σ_ = 2 the projection peaks will be close to the sampling points, so sampling errors should be much smaller. It will probably improve the rest (the deviation in your "more to the story" plots), but will not make it perfect near _r_ = 0, since this basis form cannot really reproduce a straight line.
2. Use oversampling. That is, compute the projected basis on a finer grid and then bin to integer points. 
This will approximate the integrals with a flat _h_(_x_ − _x<sub>i</sub>_) over pixels (right now _h_ is a δ function, which is not good).
3. A triangular basis. Then `Mc` is simply an identity matrix, and `M` would be much easier to compute than now. I think, the result should be close to the Dasch methods (at least, the projections are also expressed as combinations of square roots and logarithms).

@DanHickstein
> Is your correction to the basis set entirely empirical?

Partially. The modified _k_ = 0 function is the simplest analytical form that makes the basis sum "flat", so that it can approximate a flat profile in principle (the original basis could not). The projected function comes directly from the Abel integral, so the empirical part is only in the small adjustments to its samples.

PS: Regularization usually produces a much larger intensity drop near _r_ = 0, so this basis "correction" is useful only to make the "unregularized" results less miserable.

**Comment by MikhailRyazanov** [comment]
So, instead of using a "magic number", which also would not work for arbitrary _σ_, I decided to use a "universal" intensity correction. Namely, a flat distribution must be flat, with a known intensity, but BASEX gives a slightly different result. The correction profile can be taken as the ratio of that known distribution to the BASEX result, and then BASEX results for other distributions can be corrected by multiplying them with this correction profile. This does correct the overall intensity shift, the oscillations, and the _r_ = 0 artifact. Moreover, it corrects the roll-off near _r_ = 0 for reg ≠ 0.

Here are some examples:
![1d](https://user-images.githubusercontent.com/12854133/46637177-7d007300-cb4a-11e8-9931-1e4e34d41507.png)
The "corrected" looks much better than the raw output in all cases.

And some images:
![images1](https://user-images.githubusercontent.com/12854133/46637272-e8e2db80-cb4a-11e8-8690-0d39f72f3d8f.png)
![images5](https://user-images.githubusercontent.com/12854133/46637280-f4ce9d80-cb4a-11e8-8d52-b1fdb15178d9.png)
Notice how strong regularization "erases" the centerline to almost zero intensity and how "correction" brings it back, such that the rings remain continuous. (Here it also makes the axial pixels quite noisy, but I think that this is due to the strange vertical line in the experimental image; will need to try with a more reasonable data.)

**Comment by MikhailRyazanov** [comment]
But a "flat" distribution (a step function actually) has a problem that it ends too sharply, and this produces strong ringing near the outer edge, especially for broad basis sets (see top panels below).
Aligning the step edge with at a "half" of the last basis function does help in reducing this ringing  (middle panels), but the "correction profile" ("rel" in the plots) for broad basis sets is still bad near the step edge, since the step and the BASEX result have very different slopes there.
![profiles](https://user-images.githubusercontent.com/12854133/46637605-7ecb3600-cb4c-11e8-9002-50d7295d8879.png)
I tried to make a "soft step", which actually help a lot (bottom panels). I don't know any such function that has an analytical Abel transform, so this is just a piecewise polynomial (a constant 1, then a cubic, dropping from 1 to 0 with zero derivatives at both ends). The choice of the "soft edge" parameters is not scientific so far, so the plots simply show the best that I could get — it seems to work fine for any _σ_ and _n_ (even when _n_ is not a multiple of _σ_), but maybe a slightly broader edge with proper alignment will work better.

**Comment by stggh** [comment]
Dynamo @MikhailRyazanov! 
> strange vertical line in the experimental image ...  more reasonable data

`O2-ANU1024.txt` is actually an VM-image reconstructed from a single co-added quadrant, however, I do not see any vertical line:
![sliceim](https://user-images.githubusercontent.com/10932229/46639386-5de8fc00-cbb1-11e8-9fb4-504e6c7f3d92.png)


 `abel.tools.analytical.TransformPair(n, profile=5)` has a flat profile source, without soft edges. All PyAbel method struggle with it, on the outer edge, where the derivative is estimated poorly.

![tp](https://user-images.githubusercontent.com/10932229/46639461-e23b7f00-cbb1-11e8-8b4b-d03d9ea9c75e.png)

> useful to integrate that stuff into abel.tools.analytical...

Agreed

> mathematics and implementation details behind basis generation and polynomials need to be documented somewhere

`PyAbel/doc/`    as an `.rst` file.   Use  ":math:`\LaTeX`"   inline Latex and `.. math::   \LaTeX`  for multiline

Your analysis is interesting. Thanks. I may have something useful to say when I have given it more thought.

**Comment by MikhailRyazanov** [comment]
> `O2-ANU1024.txt` is actually an VM-image reconstructed from a single co-added quadrant, however, I do not see any vertical line

Here is a pixel-to-pixel rendering:
![o2-anu1024](https://user-images.githubusercontent.com/12854133/46666854-60e6eb00-cbb7-11e8-95c1-7616ae31d830.png)
I clearly see a vertical line slightly shifted to the right from the center.
And what do you mean by "reconstructed from a single co-added quadrant"? It does not look symmetric — there are two prominent blemishes on the left and several other nonsymmetric features.

**Comment by stggh** [comment]
Hmm, this must be the whole image. It has been a while since I supplied the image. For (ANU) analysis, the quadrants are usually combined, before the inverse Abel transform.

The off-centre line is from a 2 CCD sensor PCO 2000 camera.

**Comment by DanHickstein** [comment]
> Good news, everyone!

I think that we need an image of Professor Farnsworth to accompany this statement:
![image](https://user-images.githubusercontent.com/1107796/46681871-6e957400-cba9-11e8-9a44-20299cd89044.png)

> Now, in addition to making smaller basis sets by cropping larger ones, it can extend smaller basis sets to larger ones by calculating only the missing matrix elements.

That sounds pretty awesome! 

Also, your new correction factor seems to be working very well! The previous suppression of all intensity along the centerline while using the regularization factor was unsightly, and it's great that you have figured out how to overcome that! I do wonder how the angular integrations would look. It's possible that the strong suppression of the centerline intensity provides an effective way of removing the centerline noise, and if you are integrating the image anyway, it might be beneficial. In any case, we are probably better off with this more accurate transform.

I'm a bit confused how you are computing the new "magic number" (or equivalent). Do you need to recompute it for each sigma and `n`?

> which ended up as a general code for piecewise polynomial functions of any degree. I think, it would be useful to integrate that stuff into abel.tools.analytical...

sounds great!

Thanks again for all of these miraculous improvements!

**Comment by DanHickstein** [comment]
> PyAbel/doc/ as an .rst file. Use ":math:\LaTeX" inline Latex and .. math:: \LaTeX for multiline

Yes, you can make a few doc file for the polynomials, and you can make any changes to `/doc/transform_methods/basex.rst` that you would like!

**Comment by MikhailRyazanov** [comment]
> I think that we need an image of Professor Farnsworth to accompany this statement:

I had [stronger](https://img00.deviantart.net/eafe/i/2010/110/9/8/good_news_everyone_by_martynasx.jpg) implications. :–)

> I do wonder how the angular integrations would look.

It should have a sin _θ_ factor, which tends to zero at the axis, so what happens near the centerline should have very small effect on the integral results.

> I'm a bit confused how you are computing the new "magic number" (or equivalent). Do you need to recompute it for each sigma and `n`?

The correction profile takes care of the old "magic number". It is [computed](https://github.com/MikhailRyazanov/PyAbel/blob/86948f646028551d753f0c0561a729be3724fbd6/abel/basex.py#L378-L383) (if requested) immediately after computing the transform matrix and applied to it. This requires _O_(_n_<sup>2</sup>) time to compute (after _O_(_n_<sup>3</sup>) for computing the transform matrix itself) and _O_(_n_<sup>2</sup>) to apply.

An alternative would be to apply it to the transform result, which will require only _O_(_n_) for 1D transforms, but the same _O_(_n_<sup>2</sup>) overhead for square images, every time. Plus the need to pass it around.

I was also thinking about precomputing and storing the correction with the basis, but to support basis cropping, it will require correction profiles for all cropped sizes, which means doing _n_ matrix inversions. Naively, this would bring the basis-generation complexity to _O_(_n_<sup>4</sup>). The [Sherman–Morrison formula](https://en.wikipedia.org/wiki/Sherman–Morrison_formula) can help, but I don't know about the numerical stability of its repetitive applications. And this approach cannot correct the regularization effects.

**Comment by MikhailRyazanov** [comment]
>> PyAbel/doc/ as an .rst file. Use ":math:\LaTeX" inline Latex and .. math:: \LaTeX for multiline
>
> Yes, you can make a few doc file for the polynomials, and you can make any changes to /doc/transform_methods/basex.rst that you would like!

OK. Is there a good way to preview the results? (It seems that :math: is rendered only at readthedocs.io, and they use their own render, which is probably not completely compatible with arbitrary LaTeX code.)

**Comment by stggh** [comment]
> fast computer, could you please check how new basis generation

Only by chance. It is just a Dell OptiPlex 7040 that just happens to be fast. I will run the benchmarking code to cf with the current calculation.

> doc: preview the results

Follow the instructions in `CONTRIBUTING.rst` and then run `make htmllive`

**Comment by stggh** [comment]
>  check how my new basis generation works for large n? 

@MikhailRyazanov what branch should I use `basis`? Has a few bugs, because I run python3.7

1) no `sys.maxint`
2) line 350, in `get_bs_basex_cached`
    `if largest_file:`
UnboundLocalError: local variable 'largest_file' referenced before assignment

Do I need to make the changes that you are suggesting? In:
> There are two things to compare:

**Comment by MikhailRyazanov** [comment]
Oops!
Try now: https://github.com/MikhailRyazanov/PyAbel/tree/basis.

For comparing with a less restricted sum, yes, you will need to change that code as indicated. I suggest generating a basis set with the old method and saving the resulting basis file elsewhere, then doing the same with the new method, then the same with a larger `delta`. After that all 3 basis sets can be compared with each other.

**Comment by stggh** [comment]
@MikhailRyazanov I can say that your new basis calculation is much faster - nice work!

Your specific calculation request is taking a little while, as I do not have a convenient script, and I must swap versions (old `basex` vs new).  Is there a subtle approximation that occurs for each basis size, or can I just run a large basis, and then perform the smaller size comparisons by slicing the basis files, old vs new calculations?

**Comment by MikhailRyazanov** [comment]
The total size (`n`) affects only how many basis functions and how far in _r_ are calculated. Slicing larger basis files should reproduce smaller basis files exactly, so there is no need to do comparisons for different sizes. I would suggest first to test how the comparison itself works on something small and then use the largest `n` doable with the old code.

I can make the necessary scripts if you wish.

**Comment by stggh** [comment]
> I can make the necessary scripts if you wish.

Yes, please, if it is not too much trouble.

**Comment by MikhailRyazanov** [comment]
OK, here are the scripts: [gen_bs.py and cmp_bs.py](https://gist.github.com/MikhailRyazanov/90031416644f27704c85e6fadb8131e3).

Probably, the easiest way would be to:
1. Download and unpack (without installing) https://github.com/PyAbel/PyAbel/archive/master.zip and https://github.com/MikhailRyazanov/PyAbel/archive/basis.zip.
2. Copy `gen_bs.py` to `PyAbel-master/` and run it there; copy `PyAbel-master/bs_old.npy` to `PyAbel-basis/`.
3. Copy `gen_bs.py` to `PyAbel-basis/`, edit it to select the "new" method and run.
4. Edit `PyAbel-basis/abel/basex.py`, edit `PyAbel-basis/gen_bs.py` to select the "new_20" method and run (then edit `PyAbel-basis/abel/basex.py` back).
5. Copy `cmp_bs.py` to `PyAbel-basis/` and run it there, this should print the maximal deviations and generate 4 plots.
6. If everything above worked fine, then change `n` in both `gen_bs.py` files to the largest reasonable value and repeat.

**Comment by stggh** [comment]
Assuming I have correctly followed the above instructions (thanks for the scripts), here are the results for `n = 2000`. I am currently running `n = 3000` on another computer.

```
[etalon:.+st/PyAbel-basis]$ python cmp_bs.py 
new - old:
  absolute:
    max = 6.217985912826407e-07
    min = -6.465755149065444e-07
  relative:
    max = 7.704344162273468e-09
    min = -9.292370506136233e-09

new_20 - new:
  absolute:
    max = 1.786588654795196e-10
    min = -2.6645352591003757e-15
  relative:
    max = 4.479782262168209e-12
    min = -8.159725029050161e-16
```
![new - old absolute](https://user-images.githubusercontent.com/10932229/46845467-5f235e80-ce27-11e8-9507-da4d4958081f.png)
![new - old relative](https://user-images.githubusercontent.com/10932229/46845470-6185b880-ce27-11e8-8b21-f3a14d7493bb.png)
![new_20 - new absolute](https://user-images.githubusercontent.com/10932229/46845475-65b1d600-ce27-11e8-86ad-a4dbf83493cb.png)
![new_20 - new relative](https://user-images.githubusercontent.com/10932229/46845476-677b9980-ce27-11e8-94cf-05c0e33174e9.png)

**Comment by MikhailRyazanov** [comment]
Thanks!

Looks like the relative error due to the sum truncation at 7 _k_ really converges to < 5·10<sup>−12</sup> for large 'n'.
And that the new code is more accurate than the old one.

Let's see what `n = 3000` shows.
Is it the largest doable size for the old code? I thought, you had `n = 4097` (in ~4 hours at i7).

**Comment by stggh** [comment]
Running `n=5000` at the moment.  The 4 hours is cpu time(?)

```
[etalon:.+t/PyAbel-master]$ python gen_bs.py 
Generating horizontal BASEX basis sets for n = 5000:

0...50...100...150...200...250...300...350...400...450...500...550...600...650...700...750...800...850...900...950...1000...1050...1100...1150...1200...1250...1300...1350...1400...1450...1500...1550...1600...1650...1700...1750...1800...1850...1900...1950...2000...2050...2100...2150...2200...2250...2300...2350...2400...2450...2500...2550...2600...2650...2700...2750...2800...2850...2900...2950...3000...3050...3100...3150...3200...3250...3300...3350...3400...3450...3500
```
![image](https://user-images.githubusercontent.com/10932229/47050247-61046d80-d1eb-11e8-9968-1612083176e3.png)

**Comment by stggh** [comment]
Sorry this `n=5000` calculation took so long:
```
new - old:
  absolute:
    max = 8.258361134494407e-06
    min = -8.274048582279647e-06
  relative:
    max = 7.435331131939642e-08
    min = -6.331808356004442e-08

new_20 - new:
  absolute:
    max = 2.964242185043986e-10
    min = -5.329070518200751e-15
  relative:
    max = 4.701426539820991e-12
    min = -8.300373435760878e-16
```

![new_20 - new absolute](https://user-images.githubusercontent.com/10932229/47320743-0ad97380-d69e-11e8-894b-c300a77a5057.png)
![new_20 - new relative](https://user-images.githubusercontent.com/10932229/47320836-4f650f00-d69e-11e8-8816-355d4714ebfa.png)

![new - old absolute](https://user-images.githubusercontent.com/10932229/47320757-1462db80-d69e-11e8-82e6-9c7a220b1134.png)
![new - old relative](https://user-images.githubusercontent.com/10932229/47320766-175dcc00-d69e-11e8-85bf-fdd74b841da4.png)

**Comment by MikhailRyazanov** [comment]
Thanks, @stggh!

Apparently, the accuracy does not get worse for larger `n`, and the old–new discrepancy stays within ~10<sup>−8</sup>.

I have done some new-method comparisons for `n = 6000` with different summation cutoffs. It seems that at `delta = 9 * k` the results actually reach the numerical precision:
![new_20 - new_9 absolute](https://user-images.githubusercontent.com/12854133/47324925-3b46f080-d651-11e8-9bf1-0a2a4d3d97fe.png)
![new_20 - new_9 relative](https://user-images.githubusercontent.com/12854133/47324936-41d56800-d651-11e8-8a70-957c6d19d058.png)
I'm now extending this to `n = 10000` to see whether "9" is enough there as well. Then, depending on the results, I plan to change "7" to "9" or "10" — this will make the new calculations slightly (by 30–40 %) slower, although still much faster than the old method, but will eliminate that arbitrary choice of the cutoff.

Or maybe it would be better to expose this cutoff constant as a parameter, like `MAX_BASIS_SET_OFFSET` was before? Such that users can chose between the accuracy and the speed.

**Comment by stggh** [comment]
I call ~10<sup>-8</sup> agreement and cannot imagine why anyone would want machine precision for old vs new.  Expose the parameter if you wish, but set the default for maximum speed.

**Comment by MikhailRyazanov** [comment]
I also agree that 10<sup>−8</sup> is more than enough for all practical purposes, and then the cutoff can actually be reduced, making the calculations even faster.

But on the other hand, if the basis transform is claimed to be analytical, it would be nice to compute it accurately (otherwise we could use some much faster approximation).

I think, I'm going to make a table cutoff ↔ accuracy and put it in a comment (and docs), setting the default cutoff value to give that ~10<sup>−8</sup> accuracy and the highest speed.

**Comment by DanHickstein** [comment]
> I think, I'm going to make a table cutoff ↔ accuracy and put it in a comment (and docs), setting the default cutoff value to give that ~10−8 accuracy and the highest speed.

Great solution! If, in the future, someone wants to mess with this, then we can expose the parameter. Otherwise it seems like a helpful comment is the best course of action for now.

**Comment by MikhailRyazanov** [comment]
Can somebody tell me how tests work? Are they executing whatever is under `if __name__ == '__main__':` or actually calling each function with default parameters?

I want to include tests for BASEX transforms with different combinations of options, and the code is basically the same for all of them, but the parameters are different. What is the best way to do this?

**Comment by stggh** [comment]
@rth is the expert on this.

`if __name__` is ignored by testing, but useful for a local test.

AFAIK no parameters are passed to the test functions, however, inside the function you may call other functions with parameters. e.g. [see web page](https://semaphoreci.com/community/tutorials/testing-python-applications-with-pytest)

In particular:
> Pytest expects our tests to be located in files whose names begin with test_ or end with _test.py

and

```python
# test_capitalize.py

def capital_case(x):
    return x.capitalize()

def test_capital_case():
    assert capital_case('semaphore') == 'Semaphore'
```

**Comment by MikhailRyazanov** [comment]
Thanks, @stggh and @rth!

As I understand, fixtures will not work without pytest, and since we have `if __name__ ...`, it is probably better to be compatible with both modes.

But I was apparently reading something wrong and did not realize that pytest calls all and only functions with names starting with `test_`. So I've made a parameterized general function and specific `test_...` functions that call it with necessary parameters. Seems to be working.

**Comment by MikhailRyazanov** [comment]
Back to basis cutoff and accuracy. (Thanks to @stggh for the lage old–new comparison!)

Deviations in the new method between indicated cutoffs and `delta = 20 * k` for _σ_ = 1:
![rel](https://user-images.githubusercontent.com/12854133/47629990-37760b00-db36-11e8-941a-d9a49b43a44a.png)
As can be seen, 9 _k_ and 10 _k_ have the same errors, and for 15 _k_ (not shown) it is also the same, which means that the numerical precision is reached.

For broader basis functions the errors are somewhat larger, since the function shape is sampled better. Here is _σ_ = 20 to determine the upper limit:
![rel](https://user-images.githubusercontent.com/12854133/47630231-8d977e00-db37-11e8-9205-2a025c205c91.png)
I also tried _σ_ = 0.7, 2, 5, and they all have smaller errors.

So basically `delta = 6 * k` gives better than 10<sup>−8</sup> relative accuracy, and `delta = 9 * k` gives the highest achievable accuracy. I have set "6" as the default cutoff level and [exposed this parameter](https://github.com/MikhailRyazanov/PyAbel/blob/36781ea7088a4d737fa42f6cfc486e5064e4fcea/abel/basex.py#L471-L487) if somebody wants to get better accuracy at ~50% speed penalty.

**Comment by MikhailRyazanov** [comment]
And the intensity correction.

I have changed the cubic smoothstep to two parabolas and made the transition slightly broader:
![softstep](https://user-images.githubusercontent.com/12854133/47630938-84a8ab80-db3b-11e8-98bf-03fb242e41a3.png)

The ringing at the edge is now much smaller (compare with https://github.com/PyAbel/PyAbel/issues/230#issuecomment-428007005):
![cor](https://user-images.githubusercontent.com/12854133/47631012-e36e2500-db3b-11e8-9580-c0b2a0651957.png)
(Here _σ_ < 1 use reg = 1, others are without regularization.)
The left panels show the full _r_ range.
The right panels show the region near the image edge (the _r_ scale is relative to _n_ − 1) with relative and absolute differences for different image sizes _n_ not perfectly aligned with the basis functions (that is, not multiples of _σ_; the numbers in the legend indicate the _n_ differences). The last _σ_ section or _r_ is dotted, since proper input must be zero there, so relative deviations there should be not important.

Look like now the edge is treated properly in all cases, such that there is no excessive ringing, and the remaining deviations are smaller than near _r_ = 0.

**Comment by stggh** [comment]
A nice detailed analysis @MikhailRyazanov, and some fantastic improvements.  Best of all I like the new `basex` speed, a factor of  ~15x.

Time for a PR.

**Comment by MikhailRyazanov** [comment]
The overhead for using this correction is negligible for all reasonable image sizes (≳300 px), even if it is recalculated every time (when `reg` changes):
![time](https://user-images.githubusercontent.com/12854133/47631542-655f4d80-db3e-11e8-83c2-264a136e5c89.png)

_I've learned something today..._
The default BLAS in Debian is very slow. Here is the same figure using OpenBLAS:
![time](https://user-images.githubusercontent.com/12854133/47631649-cd159880-db3e-11e8-876a-74f9c1f30ec1.png)

**Comment by MikhailRyazanov** [comment]
> Time for a PR.

Yes. I have finished basic documentation update today, so as soon as #233 is merged, I'll do my PR.

**Comment by stggh** [comment]
Are you running Anaconda Python? It links to MKL:
```python
In [2]: np.show_config()
mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
blas_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
blas_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
lapack_mkl_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
lapack_opt_info:
    libraries = ['mkl_rt', 'pthread']
    library_dirs = ['/home/stg121/Software/Python/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/home/stg121/Software/Python/anaconda3/include']
```

@DanHickstein is the expert on BLAS libraries ;-). Some old (mixed) [discussion](https://github.com/PyAbel/PyAbel/issues/136#issuecomment-193283108)

My Ubuntu (now 18.10) system has openblas, but I am not 100% sure it is needed.

**Comment by MikhailRyazanov** [comment]
No, regular python stuff in Debian testing:
https://packages.debian.org/buster/python
https://packages.debian.org/buster/python-numpy
https://packages.debian.org/buster/python-scipy
Both numpy and scipy use
https://packages.debian.org/buster/libblas3
by default. Although it says "This implementation is the Fortran 77 reference implementation", so its lower performance is not a big surprise, but a more than 10-fold difference?! (The fact that it is single-threaded could give a factor of 4 at best...)

By the way. Basis calculations spend ~80% of time in `numpy.exp()`, and it apparently does not even try to use more than one core. Is there a hope that this will improve soon, or it is worth trying to use explicit multithreading in `_bs_basex()`? (I've tried a little bit, but simple Python stuff like `Pool` did not work well. In C++ it would be trivial to use OpenMP...)

**Comment by rth** [comment]
> Both numpy and scipy use packages.debian.org/buster/libblas3 by default. [..] but a more than 10-fold difference?!

Yeah, it's really slow: not threading, no SSE/AVX, worse use of L3 caches etc.

On Debian you can get OpenBlas with something like,
```
sudo apt-get remove libatlas3gf-base libatlas-dev  # or any other installed BLAS
sudo apt-get install libopenblas-dev

sudo update-alternatives  --set libblas.so.3 \
    /usr/lib/openblas-base/libopenblas.so.0
sudo update-alternatives --set liblapack.so.3 \
    /usr/lib/lapack/liblapack.so.3
```
cf http://scikit-learn.org/dev/developers/advanced_installation.html#linux (but these are outdated with respect to suggesting using ATLAS).

> As I understand, fixtures will not work without pytest, and since we have if __name__ ..., it is probably better to be compatible with both modes.

I would remove `__name__` together, there is no need to support nose -- it's a no longer maintained project that hasn't been released in 3 years.. 

> (I've tried a little bit, but simple Python stuff like Pool did not work well. In C++ it would be trivial to use OpenMP...)

You could have a try with [joblib.Parallel](https://joblib.readthedocs.io/en/latest/generated/joblib.Parallel.html#joblib.Parallel) with backend `multiprocessing` and `threading` to see if it helps,
 - The issue with multiprocessing is that it needs to serialize / deserialize objects between processes (or use mmap), and that can take longer than calculations (though it's implemented in a better way in joblib).
 - The issue with threading is that no computations can happen in parallel as long as the GIL is acquired. Here if `np.exp` releases the GIL internally (e.g. when calling some C or Cython code) it might work, but I'm not sure.

Finally, the simplest without adding a dependency might be to use OpenMP in Cython https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html (but then it's a bit more difficult to distribute).

**Comment by MikhailRyazanov** [comment]
Some good news about the summation cutoff. Making it proportional to _u_ = _x_/_σ_ instead of _k_ actually produces more uniform convergence and obviously uses less terms (since projections are non-zero only at _u_ ≲ _k_). Here is an example for _σ_ = 5 using 6 _u_:
![6u](https://user-images.githubusercontent.com/12854133/47692136-a746cd00-dbec-11e8-91ef-bbc4f087f698.png)

Computation times in seconds for different cutoffs using _u_ and _k_:

| _σ_   | _n_  | nbf  | 6 _u_ | 9 _u_ | 6 _k_ |
| :---- | ---- | ---- | ----: | ----: | ----: |
| 1     | 1000 | 1000 |    46 |    66 |    98 |
| 1     | 2000 | 2000 |   342 |   496 |   760 |
| 0.7   | 1500 | 2143 |   311 |   436 |   665 |
| 1.5   | 3000 | 2000 |   500 |   730 |  1123 |
| 5     | 5000 | 1000 |   220 |   309 |   479 |

Error bounds are the same as before.

So I'm setting the default cutoff to the "exact" 9 _u_, since it is still 1.5 times faster than the previous "~10<sup>−8</sup>" 6 _k_. If somebody wants some additional speed gain at the expense of accuracy, they can change `BASIS_SET_CUTOFF` according to its description.
